taskmgr/task_status_updater.go

```diff
@@ -0,0 +1,174 @@
+		return nil
+	}
+
+	claim := func(task *TaskRun, log *logrus.Entry) *updateError {
+		log.Info("Claiming task")
+		cr := tcclient.TaskClaimRequest{
+			WorkerGroup: ts.WorkerGroup,
+			WorkerId:    ts.WorkerId,
+		}
+		// Using the taskId and runId from the <MessageText> tag, the worker
+		// must call queue.claimTask().
+		tcrsp, callSummary, err := queue.ClaimTask(task.TaskId, strconv.FormatInt(int64(task.RunId), 10), &cr)
+		// check if an error occurred...
+		if err != nil {
+			e := &updateError{err: err.Error()}
+			// If the queue.claimTask() operation fails with a 4xx error, the
+			// worker must delete the messages from the Azure queue (except 401).
+			var errorMessage string
+			switch {
+			case callSummary.HttpResponse.StatusCode == 401 || callSummary.HttpResponse.StatusCode == 403:
+				errorMessage = "Not authorized to claim task, *not* deleting it from Azure queue!"
+			case callSummary.HttpResponse.StatusCode >= 500:
+				errorMessage = "Server error when attempting to claim task."
+			default:
+				errorMessage = "Received an error with a status code other than 401/403/500.  Deleting message from the queue"
+				// attempt to delete, but if it fails, log and continue
+				// nothing we can do, and better to return the first 4xx error
+				e.StatusCode = callSummary.HttpResponse.StatusCode
+				if err != nil {
+					errorMessage = "Not able to delete task from queue after receiving an unexpected status code."
+				}
+			}
+			log.WithFields(logrus.Fields{
+				"error":      err,
+				"statusCode": callSummary.HttpResponse.StatusCode,
+			}).Error(errorMessage)
+			return e
+		}
+		task.TaskClaimResponse = *tcrsp
+		return nil
+	}
+
+	reclaim := func(task *TaskRun, log *logrus.Entry) *updateError {
+		log.Info("Reclaiming task")
+		tcrsp, _, err := queue.ReclaimTask(task.TaskId, fmt.Sprintf("%d", task.RunId))
+
+		// check if an error occurred...
+		if err != nil {
+			return &updateError{err: err.Error()}
+		}
+
+		task.TaskReclaimResponse = *tcrsp
+		log.Info("Reclaimed task successfully")
+		return nil
+	}
+
+	go func() {
+		// only update if either IfStatusIn is nil
+		// or it is non-nil but it has "true" value
+		// for key of current status
+		if ts.IfStatusIn == nil || ts.IfStatusIn[ts.Status] {
+			task := ts.Task
+			switch ts.Status {
+			// Aborting is when you stop running a job you already claimed
+			case Succeeded:
+				e <- reportCompleted(task, logger)
+			case Failed:
+				e <- reportFailed(task, logger)
+			case Errored:
+				e <- reportException(task, ts.Reason, logger)
+			case Claimed:
+				e <- claim(task, logger)
+			case Reclaimed:
+				e <- reclaim(task, logger)
+			default:
+				err := &updateError{err: fmt.Sprintf("Internal error: unknown task status: %v", ts.Status)}
+				logger.Error(err)
+				e <- err
+			}
+		} else {
+			// current status is such that we shouldn't update to new
+			// status, so just report that no error occurred...
+			e <- nil
+		}
+	}()
+	return e
+}
```

Comment:
I think we can make the dynamic internal functions (e.g. `reportCompleted := func(...) { ... }`) exported and static, and such that instead of passing something into the update channel, you just call the method, and internally it pushes a struct down a channel. Then you would just call taskMgr.ReportCompleted(...) etc.

Also, I think the updater instance should be specific to the `TaskRun` - i.e. we could have something like:

```
	TaskRun struct {
		TaskId              string                         `json:"taskId"`
		RunId               uint                           `json:"runId"`
		SignedDeleteUrl     string                         `json:"-"`
		TaskClaimResponse   tcqueue.TaskClaimResponse      `json:"-"`
		TaskReclaimResponse tcqueue.TaskReclaimResponse    `json:"-"`
		Definition          tcqueue.TaskDefinitionResponse `json:"-"`
		reclaimTimer        *time.Timer
		taskUpdater         *taskmgr.Manager
	}
```
Then the updating of two different task runs can be done in parallel without any need for blocking, as there would be one `Manager` per task run.